### ISSUE TITLE
compileSafety: missing binding silently passes on incremental compilation - only detected with --rerun-tasks or clean

### ISSUE THREAD

**Issue by j-bajon** [InsertKoinIO/koin-compiler-plugin#32]
**Describe the bug**

The Koin Compiler Plugin's `compileSafety` graph validation does not detect a missing binding when an upstream module's `module { ... }` body is edited and the downstream consumer is re-compiled incrementally. Removing a `single<X>() bind I::class` declaration from a library module silently passes `:app:compileDebugKotlin` on a normal incremental build. The error only surfaces when the validation task is forced to fully re-execute (`--rerun-tasks`, `clean`, or a complete cache invalidation), at which point the plugin correctly reports the missing binding.

This defeats the core promise of compile-time graph validation: the build that you actually run during day-to-day development won't catch the broken graph, and the failure resurfaces only at runtime as `NoDefinitionFoundException`.

**To Reproduce**

Multi-module Android project, full DSL (no annotations):

```
:lib-api       → declares interface CatalogRepository + impl CatalogRepositoryImpl
                 declares apiClientModule = module { single<CatalogRepositoryImpl>() bind CatalogRepository::class }
                 applies io.insert-koin.compiler.plugin
:feature       → declares CatalogViewModel(catalogRepository: CatalogRepository)
                 declares catalogModule = module { viewModel<CatalogViewModel>() }
                 applies io.insert-koin.compiler.plugin, depends on :lib-api
:app           → startKoin { modules(apiClientModule, catalogModule) }
                 applies io.insert-koin.compiler.plugin, depends on :lib-api and :feature
```

Steps:
1. `./gradlew :app:compileDebugKotlin` - succeeds, graph is valid.
2. In `:lib-api`, comment out the line `single<CatalogRepositoryImpl>() bind CatalogRepository::class`.
3. `./gradlew :app:compileDebugKotlin` - **succeeds silently** (BUILD SUCCESSFUL). No `[Koin] Missing dependency` error is reported despite `CatalogViewModel` requiring `CatalogRepository`.
4. `./gradlew :app:compileDebugKotlin --rerun-tasks` (or `:app:clean` first) - **correctly fails** with:
   ```
   e: [Koin] Missing dependency: com.example.lib.CatalogRepository
     required by: dsl:com.example.feature.CatalogViewModel (parameter 'catalogRepository')
     in module: DSL graph
   ```
5. Without the workaround, the broken graph reaches the device and crashes at runtime:
   ```
   org.koin.core.error.NoDefinitionFoundException: No definition found for type
     'com.example.lib.CatalogRepository' on scope '['_root_']'
   ```

**Expected behavior**

Step 3 should fail with the same `[Koin] Missing dependency` error as step 4. The plugin's task inputs (or its generated hint output declarations) should be wired so that any change to a `module { ... }` body in the transitive Koin graph invalidates the up-to-date check on `:app:compileDebugKotlin`, ensuring the full graph is re-validated even on incremental builds.

In practice, what seems to happen is: editing only the body of `val apiClientModule = module { ... }` does not change the public ABI of `:lib-api`, so Kotlin's incremental compilation does not invalidate `:app:compileDebugKotlin`. The plugin then validates a partial graph (only the files Kotlin actually re-compiled in `:app`) and the missing binding goes unnoticed.

A user-level workaround exists (`--rerun-tasks` in CI or `outputs.upToDateWhen { false }` on `compileDebugKotlin`), but both defeat incremental compilation and shift the burden onto consumers. Plugin-level options that would solve it cleanly:
- Track every Koin source file (any file containing `module { }`, `single<>`, `viewModel<>`, etc.) in the transitive classpath as an explicit Gradle task input, so body-level edits invalidate the validation.
- Emit a "graph fingerprint" output that participates in Kotlin's classpath snapshot so body edits surface as ABI-relevant for the plugin.
- Provide a separate, non-incremental `koinValidate` task that always re-runs the full graph check.

**Koin module and version:**

- `io.insert-koin:koin-bom:4.2.1`
- `io.insert-koin.compiler.plugin:1.0.0-RC2`
- Kotlin `2.3.20`
- AGP `9.2.0`
- Gradle `9.4.1`
- Plugin defaults (no `koinCompiler {}` block): `compileSafety=true`, `unsafeDslChecks=true`, `skipDefaultValues=true`

**Snippet or Sample project to help reproduce**

Minimal snippets (full Android project not needed - same behavior should reproduce on any multi-module Kotlin project where the plugin is applied to the aggregating module):

```kotlin
// :lib-api/src/main/kotlin/.../CatalogRepository.kt
interface CatalogRepository
class CatalogRepositoryImpl : CatalogRepository

// :lib-api/src/main/kotlin/.../ApiClientModule.kt
val apiClientModule = module {
    single<CatalogRepositoryImpl>() bind CatalogRepository::class // ← comment this line
}

// :feature/src/main/kotlin/.../CatalogViewModel.kt
class CatalogViewModel(
    private val catalogRepository: CatalogRepository,
) : ViewModel()

// :feature/src/main/kotlin/.../CatalogModule.kt
val catalogModule = module {
    viewModel<CatalogViewModel>()
}

// :app/src/main/kotlin/.../Application.kt
startKoin {
    modules(apiClientModule, catalogModule)
}
```

Happy to put together a tiny repo if helpful - just let me know.

**Comment by arnaudgiuliani** [comment]
Hey Julien, thanks for the very thorough write-up, your diagnosis is exactly right.

The root cause is fundamental to Kotlin's K2 incremental compilation model: when the aggregator module (:app) has no source file that directly references the declaration that changed, its compileKotlin task stays UP-TO-DATE and the IR safety pass never runs. LookupTracker / ExpectActualTracker only works within a module; cross-module IC uses ABI snapshots, and a DSL binding lives in a lambda body, which isn't ABI. Metro hits the same wall, it just dodges it by being annotation-only.

Two fixes landed on main and will ship in the next release:

  1. `koinCompiler { strictSafety = true }` — opt-in flag (default false). Set on the module containing startKoin { } / @KoinApplication; that task always re-runs the safety pass. Other modules stay fully incremental. Cost = roughly that one task's compile time per build, which on a thin :app is small.
  2. Stable synthetic hint file anchors — the dsl_* / callsite(...) / injectedparams_* hints now anchor on the DSL call's own source file (or sorted-deterministic fallback) instead of an arbitrary module file. Kotlin's IC can now track them across incremental rebuilds. This addresses the residual stale-output mode.

  If you want to try it early, I can publish a snapshot you can pin against. A standalone koinValidate task that bypasses compileKotlin entirely is on the roadmap as a follow-up.

Should be good now, once you specify `koinCompiler { strictSafety = true }` in your main Koin DSL module.

**Comment by arnaudgiuliani** [comment]
Fixed in 91a9b20e088b4a081d96c28f0887a837bf209a13
